### ISSUE TITLE
Cannot use Argo CD clientset with K8s 1.28

### ISSUE THREAD

**Issue by Bondza** [argoproj/argo-cd#17935]
Checklist:

* [x] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [x] I've included steps to reproduce the bug.
* [x] I've pasted the output of `argocd version`.

**Describe the bug**

We use the Argo CD Go [library](https://pkg.go.dev/github.com/argoproj/argo-cd/v2) and generated clientset to manage Applications in our cluster. When trying to update to K8s dependencies like `k8s.io/kubectl` to version v0.28.8 it fails to build because a function has changed signature:

```
# github.com/argoproj/gitops-engine/pkg/utils/kube
/home/codespace/go/pkg/mod/github.com/argoproj/gitops-engine@v0.7.1-0.20240122213038-792124280fcc/pkg/utils/kube/resource_ops.go:269:39: too many arguments in call to apply.NewApplyFlags
	have ("k8s.io/kubectl/pkg/cmd/util".Factory, genericiooptions.IOStreams)
	want (genericiooptions.IOStreams)
/home/codespace/go/pkg/mod/github.com/argoproj/gitops-engine@v0.7.1-0.20240122213038-792124280fcc/pkg/utils/kube/resource_ops.go:272:22: cannot use sets.NewString() (value of type "k8s.io/apimachinery/pkg/util/sets".String) as "k8s.io/apimachinery/pkg/util/sets".Set["k8s.io/apimachinery/pkg/types".UID] value in struct literal
/home/codespace/go/pkg/mod/github.com/argoproj/gitops-engine@v0.7.1-0.20240122213038-792124280fcc/pkg/utils/kube/resource_ops.go:273:22: cannot use sets.NewString() (value of type "k8s.io/apimachinery/pkg/util/sets".String) as "k8s.io/apimachinery/pkg/util/sets".Set[string] value in struct literal
/home/codespace/go/pkg/mod/github.com/argoproj/gitops-engine@v0.7.1-0.20240122213038-792124280fcc/pkg/utils/kube/resource_ops.go:290:4: o.DryRunVerifier undefined (type *apply.ApplyOptions has no field or method DryRunVerifier)
/home/codespace/go/pkg/mod/github.com/argoproj/gitops-engine@v0.7.1-0.20240122213038-792124280fcc/pkg/utils/kube/resource_ops.go:290:74: k.fact.OpenAPIGetter undefined (type "k8s.io/kubectl/pkg/cmd/util".Factory has no field or method OpenAPIGetter)
/home/codespace/go/pkg/mod/github.com/argoproj/gitops-engine@v0.7.1-0.20240122213038-792124280fcc/pkg/utils/kube/resource_ops.go:290:100: undefined: resource.QueryParamDryRun
/home/codespace/go/pkg/mod/github.com/argoproj/gitops-engine@v0.7.1-0.20240122213038-792124280fcc/pkg/utils/kube/resource_ops.go:291:4: o.FieldValidationVerifier undefined (type *apply.ApplyOptions has no field or method FieldValidationVerifier)
/home/codespace/go/pkg/mod/github.com/argoproj/gitops-engine@v0.7.1-0.20240122213038-792124280fcc/pkg/utils/kube/resource_ops.go:291:83: k.fact.OpenAPIGetter undefined (type "k8s.io/kubectl/pkg/cmd/util".Factory has no field or method OpenAPIGetter)
/home/codespace/go/pkg/mod/github.com/argoproj/gitops-engine@v0.7.1-0.20240122213038-792124280fcc/pkg/utils/kube/resource_ops.go:296:59: o.DryRunVerifier undefined (type *apply.ApplyOptions has no field or method DryRunVerifier)
/home/codespace/go/pkg/mod/github.com/argoproj/gitops-engine@v0.7.1-0.20240122213038-792124280fcc/pkg/utils/kube/resource_ops.go:360:74: k.fact.OpenAPIGetter undefined (type "k8s.io/kubectl/pkg/cmd/util".Factory has no field or method OpenAPIGetter)
/home/codespace/go/pkg/mod/github.com/argoproj/gitops-engine@v0.7.1-0.20240122213038-792124280fcc/pkg/utils/kube/resource_ops.go:360:74: too many errors
```

The [version skew policy](https://kubernetes.io/releases/version-skew-policy/#kubectl) for `kubectl` states:
> kubectl is supported within one minor version (older or newer) of kube-apiserver

And Argo CD is on [v0.26.4](https://github.com/argoproj/argo-cd/blob/master/go.mod#L103) for `k8s.io/kubectl`, so 1.28 for example is not officially supported with the current version.

**To Reproduce**
 
Try to build an application with these dependencies in `go.mod`:
```
	github.com/argoproj/argo-cd/v2 v2.10.1
	k8s.io/kubectl v0.28.8
```

**Expected behavior**

It builds successfully.

**Screenshots**

N/A

**Version**

```shell
argocd: v2.10.1+a79e0ea
  BuildDate: 2024-02-14T17:56:39Z
  GitCommit: a79e0eaca415461dc36615470cecc25d6d38cefb
  GitTreeState: clean
  GoVersion: go1.21.7
  Compiler: gc
  Platform: linux/amd64
argocd-server: v2.10.1+a79e0ea
```

**Logs**

N/A

**Comment by jannfis** [comment]
Unfortunately, you need to pin the Kubernetes version in your consumer to the same version that Argo CD uses. This is done using replace rules in your go.mod.

PTAL here: https://github.com/argoproj/argo-cd/blob/b060053b099b4c81c1e635839a309c9c8c1863e9/go.mod#L310-L338

**Comment by Bondza** [comment]
We do pin the Kubernetes version at the moment and this works since we use v0.26.10 today. But we are moving to Kubernetes 1.28 and this is why we would like to update the version.

The issue is that we have other dependencies to update but they require newer versions of Kubernetes. Pinning the version means that we are not able to update other dependencies that require a newer version of the Kubernetes dependencies.

Is there a plan to update the Kubernetes version that Argo CD uses?

**Comment by joselitofilho** [comment]
I am moving to Kubernetes 1.28/1.29. So I would also like to update the Argo CD version.

I have the same question as @Bondza:
> Is there a plan to update the Kubernetes version that Argo CD uses?

**Comment by blakepettersson** [comment]
There's an going effort in #17491, but seems to have stalled for now.

**Comment by sivchari** [comment]
@Bondza @joselitofilho @blakepettersson 
Hi all.
I'm sorry that I was late.
I'd work on left task. Only what I do is to pass the CI 👍

**Comment by blakepettersson** [comment]
@sivchari no worries, there is no deadline. You have been doing a great job so far, keep it up! 😄

**Comment by joselitofilho** [comment]
@sivchari no worries. Thank you for your initiative and great work! 👏

**Comment by sivchari** [comment]
I've completed all required tasks. So It's ready to be merged. Thanks.

**Comment by blakepettersson** [comment]
Fixed with #17491
